platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/ResolutionFailureHandler.java

```diff
@@ -209,6 +219,25 @@ private ComponentIdentifier getOrCreateVariantSetComponentIdentifier(ResolvedVar
     }
     // endregion Artifact Selection failures
 
+    private static ImmutableCapabilities realizeCapabilitySelectors(ComponentGraphResolveState targetComponent, Set<CapabilitySelector> requestedCapabilities) {
+        Set<ImmutableCapability> realized = new LinkedHashSet<>();
+        for (CapabilitySelector selector : requestedCapabilities) {
+            if (selector instanceof SpecificCapabilitySelector) {
+                DefaultSpecificCapabilitySelector specificSelector = (DefaultSpecificCapabilitySelector) selector;
+                realized.add(new DefaultImmutableCapability(specificSelector.getGroup(), specificSelector.getName(), null));
+            } else if (selector instanceof FeatureCapabilitySelector) {
+                FeatureCapabilitySelector featureSelector = (FeatureCapabilitySelector) selector;
+                ModuleVersionIdentifier moduleId = targetComponent.getMetadata().getModuleVersionId();
+                realized.add(new DefaultImmutableCapability(
+                    moduleId.getGroup(),
+                    moduleId.getName() + "-" + featureSelector.getFeatureName(),
+                    null
+                ));
+            }
```

Comment:
❓ I'm not sure I see the point of this conversion.  Why can't we just adapt the few failure types that deal with capabilities to use `Set<CapabilitySelector>`?  

They don't see like particularly heavyweight types and are already immutable.

We should probably adapt the failure message generators to work off of them directly as well.

Also, if we do leave it this way, which isn't my first choice, it seems like the `CapabilitySelector` interface needs a `toCapability(<target component type>)` method.  

Alternately, could this be a static factory conversion method that lives in `CapabilitySelector` itself and adds an `else`?  Would be easier to keep it in sync with any new types and not have not really cohesive type conversation code live in this handler class.

Reply:
This PR is already at 1200+ lines and 100+ files. I thought this was a good point to hold off on further changes as your proposed changes would just widen the scope even more here. 

I agree the failure types should operate on capability selectors instead of requested capabilities, so this code here was just to reduce the scope of these changes where possible. 

I can implement those changes -- its just a question if its worth doing it at the moment or not, as it means a non insignificant chunk of changes to add to this PR. 

Up to you if you if you want me to add them to this PR

Reply:
I suppose it's not critical.  Better to split into a new issue, that should be a pretty quick PR.

Reply:
I will write an issue for this

Reply:
https://github.com/gradle/gradle/issues/30549